src/vmm/src/devices/virtio/net/device.rs

```diff
@@ -312,25 +363,381 @@ impl Net {
+    }
+
+    fn process_tx(&mut self) -> Result<(), DeviceError> {
+        if !self.readiness.tx_ready() {
+            return Ok(());
+        }
+
+        // This is safe since we checked in the event handler that the device is activated.
+        let mem = self.device_state.mem().unwrap();
+
+        // The MMDS network stack works like a state machine, based on synchronous calls, and
+        // without being added to any event loop. If any frame is accepted by the MMDS, we also
+        // trigger a process_rx() which checks if there are any new frames to be sent, starting
+        // with the MMDS network stack.
+        let mut process_rx_for_mmds = false;
+        let mut used_any = false;
+        let tx_queue = &mut self.queues[TX_INDEX];
+
+        loop {
+            let Some(head) = tx_queue.pop_or_enable_notification(mem) else {
+                // Queue is empty now, clear readiness and leave.
+                self.readiness.clear_tx_ready();
+                break;
+            };
+            self.metrics
+                .tx_remaining_reqs_count
+                .add(tx_queue.len(mem).into());
+            let head_index = head.index;
+
+            macro_rules! push_desc {
+                () => {
+                    tx_queue
+                        .add_used(mem, head_index, 0)
+                        .map_err(DeviceError::QueueError)?;
+                    used_any = true;
+                };
+            }
+
+            // Parse IoVecBuffer from descriptor head
+            // SAFETY: This descriptor chain is only loaded once
+            // virtio requests are handled sequentially so no two IoVecBuffers
+            // are live at the same time, meaning this has exclusive ownership over the memory
+            if unsafe { self.tx_buffer.load_descriptor_chain(head).is_err() } {
+                self.metrics.tx_fails.inc();
+                push_desc!();
+                continue;
+            };
+
+            // We only handle frames that are up to MAX_BUFFER_SIZE
+            if self.tx_buffer.len() as usize > MAX_BUFFER_SIZE {
+                error!("net: received too big frame from driver");
+                self.metrics.tx_malformed_frames.inc();
+                push_desc!();
+                continue;
+            }
+
+            if !Self::rate_limiter_consume_op(
+                &mut self.tx_rate_limiter,
+                u64::from(self.tx_buffer.len()),
+            ) {
+                tx_queue.undo_pop();
+                self.metrics.tx_rate_limiter_throttled.inc();
+                break;
+            }
+
+            let frame_consumed_by_mmds = Self::write_to_mmds_or_tap(
+                self.mmds_ns.as_mut(),
+                &mut self.tx_rate_limiter,
+                &mut self.tx_frame_headers,
+                &self.tx_buffer,
+                &mut self.tap,
+                self.guest_mac,
+                &self.metrics,
```

Comment:
Would be nice if you could make this function work on `self` (so `self.write_to_mmds_or_tap`). It only takes `self.*` arguments anyway.